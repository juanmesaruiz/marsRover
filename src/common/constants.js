export const ROVER_DIRECTION = {
  N: 'north',
  S: 'south',
  E: 'east',
  W: 'west',
};

export const ROVER_MOVEMENT = {
  F: 'front',
  L: 'left',
  R: 'right',
};

export const KEYBOARDS_CODES = {
  A: 'KeyA',
  D: 'KeyD',
  W: 'KeyW',
};
