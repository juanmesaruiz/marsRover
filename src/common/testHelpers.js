import React from 'react';

export const mockComponent = (name, props) => {
  const { children, ...restOfProps } = props;
  return (
    <code name={name} aria-label={name} props={restOfProps}>
      {children}
    </code>
  );
};
