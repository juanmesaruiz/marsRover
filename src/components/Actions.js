import React from 'react';
import styled from 'styled-components';

import InstructionsForm from './InstructionsForm';
import RandomObstacleButton from './RandomObstacleButton';
import LogObstacles from './LogObstacles';

const Actions = ({ className }) => (
  <div className={className}>
    <InstructionsForm />
    <RandomObstacleButton />
    <LogObstacles />
  </div>
);

const styledActions = styled(Actions)`
  display: inline-block;
  height: 376px;
  overflow: hidden;
  border: 1px solid black;
  margin-left: 15px;
  padding: 20px 20px;
  width: 200px;
`;

export default styledActions;
