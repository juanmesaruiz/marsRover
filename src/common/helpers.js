import React from 'react';
import { KEYBOARDS_CODES, ROVER_DIRECTION, ROVER_MOVEMENT } from './constants';

const isCoordinateInsideGrid = (grid, newPosition) =>
  newPosition.x >= 0 &&
  newPosition.x <= grid.x &&
  newPosition.y >= 0 &&
  newPosition.y <= grid.y;

const isObstacleCoordinates = (obstaclesCoordinates, newPosition) =>
  obstaclesCoordinates.findIndex(
    obstacle => obstacle.x === newPosition.x && obstacle.y === newPosition.y
  ) !== -1;

export const getDirectionArrow = direction => {
  switch (direction) {
    case ROVER_DIRECTION.W:
      return <span>&#8592;</span>;

    case ROVER_DIRECTION.E:
      return <span>&#8594;</span>;

    case ROVER_DIRECTION.S:
      return <span>&#8595;</span>;

    case ROVER_DIRECTION.N:
    default:
      return <span>&#8593;</span>;
  }
};

export const getNewObstaclesCoordinates = (
  grid,
  roverPosition,
  obstaclesCoordinates
) => {
  const newObstacleCoordinate = getRandomRoverPosition(
    grid,
    obstaclesCoordinates
  );
  const isRoverPosition =
    newObstacleCoordinate.x === roverPosition.x &&
    newObstacleCoordinate.y === roverPosition.y;

  if (isRoverPosition) {
    return getNewObstaclesCoordinates(
      grid,
      roverPosition,
      obstaclesCoordinates
    );
  }

  return [newObstacleCoordinate];
};

export const getRandomCoordinates = ({ x, y }, numberOfObstacles = 1) => {
  const coordinates = [];
  let i = 0;

  for (i; i < numberOfObstacles; i++) {
    const newX = Math.floor(Math.random() * (x - 0));
    const newY = Math.floor(Math.random() * (y - 0));
    const existingObstacle = coordinates.findIndex(
      obstacle => obstacle.x === newX && obstacle.y === newY
    );

    if (existingObstacle === -1) {
      coordinates.push({
        x: newX,
        y: newY,
      });
    } else {
      numberOfObstacles = numberOfObstacles + 1;
    }
  }

  return [...new Set(coordinates)];
};

export const getRandomRoverPosition = (grid, obstaclesCoordinates) => {
  const roverPosition = getRandomCoordinates(grid)[0];
  const isObstacle = isObstacleCoordinates(obstaclesCoordinates, roverPosition);

  if (isObstacle) {
    return getRandomRoverPosition(grid, obstaclesCoordinates);
  }

  return roverPosition;
};

export const getRoverMovementFromCode = code => {
  switch (code) {
    case KEYBOARDS_CODES.A:
      return ROVER_MOVEMENT.L;
    case KEYBOARDS_CODES.D:
      return ROVER_MOVEMENT.R;
    case KEYBOARDS_CODES.W:
      return ROVER_MOVEMENT.F;
    default:
      return null;
  }
};

export const isCorrectMovement = ({
  obstaclesCoordinates,
  newPosition,
  grid,
}) => {
  const isObstacle = isObstacleCoordinates(obstaclesCoordinates, newPosition);
  const isInsideGrid = isCoordinateInsideGrid(grid, newPosition);

  return !isObstacle && isInsideGrid;
};
