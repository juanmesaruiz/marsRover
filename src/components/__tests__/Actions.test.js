import React from 'react';
import ReactDOM from 'react-dom';
import { render } from '@testing-library/react';
import 'jest-styled-components';

import { mockComponent } from '../../common/testHelpers';

import Actions from '../Actions';

jest.mock('../InstructionsForm', () => props =>
  mockComponent('InstructionsForm', props)
);
jest.mock('../RandomObstacleButton', () => props =>
  mockComponent('RandomObstacleButton', props)
);
jest.mock('../LogObstacles', () => props =>
  mockComponent('LogObstacles', props)
);

describe('Actions test suite', () => {
  const component = <Actions />;

  const setup = () => render(component);

  it('Renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(component, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('Should match snapshot', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });
});
