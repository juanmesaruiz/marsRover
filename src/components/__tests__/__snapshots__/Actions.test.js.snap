// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions test suite Should match snapshot 1`] = `
.c0 {
  display: inline-block;
  height: 376px;
  overflow: hidden;
  border: 1px solid black;
  margin-left: 15px;
  padding: 20px 20px;
  width: 200px;
}

<div>
  <div
    class="c0"
  >
    <code
      aria-label="InstructionsForm"
      name="InstructionsForm"
      props="[object Object]"
    />
    <code
      aria-label="RandomObstacleButton"
      name="RandomObstacleButton"
      props="[object Object]"
    />
    <code
      aria-label="LogObstacles"
      name="LogObstacles"
      props="[object Object]"
    />
  </div>
</div>
`;
